ValueEncoder protocol

IceDragon200 opened the pull request:
Sometimes you forget to cast a value to a string.

This PR adds a `Logfmt.ValueEncoder` protocol as well as refactoring the Encoder module.

## Changes

* Added `Logfmt.ValueEncoder` protocol
* Added ValueEncoder implementations for ~~NaiveDateTime~~, ~~DateTime~~, PID, and Reference
* Added section to README introducing the ValueEncoder protocol

## Considerations

* ~~Maybe removing the NaiveDateTime and DateTime implementations would be better, have the user implement it themselves if they need a different format for datetimes.~~ NaiveDateTime and DateTime have been removed in favour of user implementations.

* Passing options from the Logfmt.Encoder.encode to the Logfmt.ValueEncoder.encode, allowing the configuration of the protocol (especially for NaiveDateTime and DateTime)

* Using `encode_value` on the key to convert atoms into strings, maybe skip that and go straight for `Logfmt.ValueEncoder.encode`

jclem commented on lib/logfmt/encoder.ex:
```diff
@@ -14,40 +51,38 @@ defmodule Logfmt.Encoder do
-    else
-      ~s(#{key}=#{encode_value(value)})
+  @spec encode(Dict.t, options :: Keyword.t) :: String.t
+  def encode(list, options \\ []) do
```
This option should also be exposed and documented in `Logfmt.encode/1` (now `Logfmt.encode/2`).

IceDragon200 replied:
Added small example of the iolist output

jclem commented on lib/logfmt/encoder.ex:
```diff
@@ -14,40 +51,38 @@ defmodule Logfmt.Encoder do
+  defp encode_pair({key, value}, []) do
+    # this is called for the first item in the list, since the accumulator
+    # would be empty at that point.
+    [[encode_value(key), "=", encode_value(value)]]
```
This isn't actually necessary—it results in something that looks like this:

```elixir
iex> Logfmt.Encoder.encode([foo: 1, bar: 2], output: :iolist)
[[["foo", "=", "1"]], " ", ["bar" ,"=", "2"]]
```

Unwrap the list, and you'll correctly get:

```elixir
iex> Logfmt.Encoder.encode([foo: 1, bar: 2], output: :iolist)
[["foo", "=", "1"], " ", ["bar" ,"=", "2"]]
```

The `output: :iolist` should have a test case, too.

IceDragon200 replied:
Ah, yes originally I was using the list join `|`, hence the double wrapped list, but I threw out the reverse code and never changed it back.

jclem commented on lib/logfmt/encoder.ex:
```diff
@@ -1,3 +1,40 @@
+defprotocol Logfmt.ValueEncoder do
```
A lot of these types already implement the `String.Chars` protocol, so I think it's fine to just return the value rather than going through the work of stringifying them here.

jclem replied:
Also, can we put the protocol in its own file and document it? I find it distracting if I'm trying to look at the encoder. Protocols in the Elixir source code [are their own files](https://github.com/elixir-lang/elixir/blob/1406d853e0e6515007696f871d0a9e2c023da6da/lib/elixir/lib/string/chars.ex#L3), as well.

IceDragon200 replied:
Regarding String.Chars

Enum.join will fail given a iolist, since it has mixed data and contain nested lists.

It works at the top-level

```elixir
Enum.join [:atom, " was an atom while ", 12, " was an integer"]
# => "atom was an atom while 12 was an integer"
```

But once you enter the world of iolists or list of lists...

```
iex(2)> Enum.join [[:atom, " was an atom while "], [12, " was an integer"]]
** (ArgumentError) cannot convert the given list to a string.

To be converted to a string, a list must contain only:

  * strings
  * integers representing Unicode codepoints
  * or a list containing one of these three elements

Please check the given list or call inspect/1 to get the list representation, got:

[:atom, " was an atom while "]

    (elixir) lib/list.ex:742: List.to_string/1
    (elixir) lib/enum.ex:1811: Enum."-join/2-lists^foldl/2-0-"/3
    (elixir) lib/enum.ex:1811: Enum.join/2
```

jclem commented on lib/logfmt/encoder.ex:
```diff
@@ -14,40 +51,38 @@ defmodule Logfmt.Encoder do
+    cond do
+      # originally I wanted to check if the string contained both ' and "
+      # the quoting would require escaping those characters first
+      String.contains?(str, " ") or String.contains?(str, "=") ->
```
Should probably check for any whitespace, right?

```elixir
String.match?(str, ~r/\s/) or String.contains?(str, "=")
```

IceDragon200 replied:
Done!

jclem commented on lib/logfmt/encoder.ex:
```diff
@@ -14,40 +51,38 @@ defmodule Logfmt.Encoder do
+  @spec encode_value(value :: term) :: String.t
+  defp encode_value(value) do
+    str = Logfmt.ValueEncoder.encode(value)
+    cond do
```
There are only two branches here, mind refactoring it to an `if`?

IceDragon200 replied:
I was using cond since I intended to check for double or single quotes in the string before wrapping.

I've changed it to an if now

jclem commented on lib/logfmt/encoder.ex:
```diff
@@ -14,40 +51,38 @@ defmodule Logfmt.Encoder do
-      ~s(#{key}=#{encode_value(value)})
+  @spec encode(Dict.t, options :: Keyword.t) :: String.t
+  def encode(list, options \\ []) do
+    result = Enum.reduce(list, [], &encode_pair/2)
```
Technically the reduce may be more efficient, but I think this may be a little more readable:

```
result =
  list
  |> Enum.map(&encode_pair/1)
  |> Enum.intersperse(" ")
```

Which also makes the `encode_pair/1` function not have to be aware of the accumulator, which is nice.

IceDragon200 replied:
And done, that was pretty neat, I've never used intersperse before

jclem commented:
Note that this PR also requires bumping the minimum Elixir version to "1.4", which I'm okay with. This is because of the use of `DateTime.from_iso8601`. Please do that, as well!

jclem commented:
Although now that I think about it, I think you're right about the DateTime stuff being better left up to the user (it would be great to have that as an example in the Logfmt.ValueEncoder docs). That way, we can keep the Elixir version required down low, still.

IceDragon200 commented:
@jclem I've made most of the suggested changes, the only caveat is the String.Chars one which I've explained.

Overall:
* Removed DateTime and NaiveDateTime implementations including tests
* Included aforementioned in the Readme as an example
* Updated Logfmt.encode to accept the options list making it Logfmt.encode/2
* Moved ValueEncoder to it's own file
* Added Encoder.encode/2 iolist test

jclem commented:
Thanks, @IceDragon200! This is now released in v3.3.0.
